From 042adc371309b726f8c47b4d8ee5b71c05fea351 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:04:28 -0400
Subject: [kernel] perf: Simplify perf_event_exit_task_context()

Message-id: <1410181752-25631-24-git-send-email-jolsa@redhat.com>
Patchwork-id: 91722
O-Subject: [PATCH RHEL7.1 BZ1134356 023/307] perf: Simplify perf_event_exit_task_context()
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit 3a497f48637e2aac17eabb84a17f8ac5216028fc
Author: Peter Zijlstra <peterz@infradead.org>
Date: Mon May 5 12:42:29 2014 +0200

description
===========
Instead of jumping through hoops to make sure to find (and exit) each
event, do it the simple straight fwd way.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/events/core.c b/kernel/events/core.c
index f7ce810..1ce2fca 100644
--- a/kernel/events/core.c
+++ b/kernel/events/core.c
@@ -7501,24 +7501,9 @@ static void perf_event_exit_task_context(struct task_struct *child, int ctxn)
   */
  mutex_lock(&child_ctx->mutex);
 
-again:
- list_for_each_entry_safe(child_event, tmp, &child_ctx->pinned_groups,
-     group_entry)
-  __perf_event_exit_task(child_event, child_ctx, child);
-
- list_for_each_entry_safe(child_event, tmp, &child_ctx->flexible_groups,
-     group_entry)
+ list_for_each_entry_rcu(child_event, &child_ctx->event_list, event_entry)
   __perf_event_exit_task(child_event, child_ctx, child);
 
- /*
-  * If the last event was a group event, it will have appended all
-  * its siblings to the list, but we obtained 'tmp' before that which
-  * will still point to the list head terminating the iteration.
-  */
- if (!list_empty(&child_ctx->pinned_groups) ||
-     !list_empty(&child_ctx->flexible_groups))
-  goto again;
-
  mutex_unlock(&child_ctx->mutex);
 
  put_ctx(child_ctx);
-- 
1.7.1